runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu

```diff
@@ -194,39 +261,99 @@ protected:
+
+    // Check if this is the first time to allocate, if so
+    // the allocation is much easier
     if (!deviceStateVector) {
+      // Create the memory and the handle
       HANDLE_CUDA_ERROR(cudaMalloc((void **)&deviceStateVector,
                                    stateDimension * sizeof(CudaDataType)));
-      constexpr int32_t threads_per_block = 256;
-      uint32_t n_blocks =
-          (stateDimension + threads_per_block - 1) / threads_per_block;
-      initializeDeviceStateVector<<<n_blocks, threads_per_block>>>(
-          reinterpret_cast<CudaDataType *>(deviceStateVector), stateDimension);
       HANDLE_ERROR(custatevecCreate(&handle));
+
+      // If no state provided, initialize to the zero state
+      if (state == nullptr) {
+        initializeDeviceStateVector<<<n_blocks, threads_per_block>>>(
+            reinterpret_cast<CudaDataType *>(deviceStateVector),
+            stateDimension);
+        return;
+      }
+
+      // User state provided...
+
+      // FIXME handle case where pointer is a device pointer
+
+      // First allocation, so just set the user provided data here
+      HANDLE_CUDA_ERROR(cudaMemcpy(deviceStateVector, state,
+                                   stateDimension * sizeof(CudaDataType),
+                                   cudaMemcpyHostToDevice));
+      return;
+    }
+
+    // State already exists, need to allocate new state and compute
+    // kronecker product with existing state
+
+    // Allocate new vector to place the kron prod result
+    void *newDeviceStateVector;
+    HANDLE_CUDA_ERROR(cudaMalloc((void **)&newDeviceStateVector,
+                                 stateDimension * sizeof(CudaDataType)));
+
+    // Place the state data on device. Could be that
+    // we just need the zero state, or the user could have provided one
+    void *otherState;
+    HANDLE_CUDA_ERROR(cudaMalloc((void **)&otherState,
+                                 stateDimension * sizeof(CudaDataType)));
+    if (state == nullptr) {
+      initializeDeviceStateVector<<<n_blocks, threads_per_block>>>(
+          reinterpret_cast<CudaDataType *>(otherState), (1UL << count));
     } else {
-      // Allocate new state..
-      void *newDeviceStateVector;
-      HANDLE_CUDA_ERROR(cudaMalloc((void **)&newDeviceStateVector,
-                                   stateDimension * sizeof(CudaDataType)));
-      constexpr int32_t threads_per_block = 256;
-      uint32_t n_blocks =
-          (stateDimension + threads_per_block - 1) / threads_per_block;
-      setFirstNElements<<<n_blocks, threads_per_block>>>(
-          reinterpret_cast<CudaDataType *>(newDeviceStateVector),
-          reinterpret_cast<CudaDataType *>(deviceStateVector),
-          previousStateDimension);
-      HANDLE_CUDA_ERROR(cudaFree(deviceStateVector));
-      deviceStateVector = newDeviceStateVector;
+
+      // FIXME Handle case where data is already on GPU
+      std::vector<std::complex<ScalarType>> in(state, state + stateDimension);
```

Comment:
I think we might be accessing the input `state` buffer beyond its memory region here.